torch_audiomentations/core/transforms_interface.py

```diff
@@ -13,45 +14,87 @@ class EmptyPathException(Exception):
+        super(BaseWaveformTransform, self).__init__()
+        assert 0.0 <= p <= 1.0
+        self._p = p
+        self.parameters = {}
         self.are_parameters_frozen = False
-
-    def forward(self, samples, sample_rate):
-        if not self.are_parameters_frozen:
-            self.randomize_parameters(samples, sample_rate)
-
-        if self.parameters["should_apply"] and len(samples) > 0:
-            if is_multichannel(samples):
-                if samples.shape[1] > samples.shape[2]:
-                    warnings.warn(
-                        "Multichannel audio must have channels first, not channels last. In"
-                        " other words, the shape must be (batch size, channels, samples), not"
-                        " (batch_size, samples, channels)"
-                    )
-                if not self.supports_multichannel:
-                    raise MultichannelAudioNotSupportedException(
-                        "{} only supports mono audio, not multichannel audio".format(
-                            self.__class__.__name__
-                        )
+        self.bernoulli_distribution = Bernoulli(self._p)
+
+    @property
+    def p(self):
+        return self._p
+
+    @p.setter
+    def p(self, p):
+        self._p = p
+        # Update the Bernoulli distribution accordingly
+        self.bernoulli_distribution = Bernoulli(self._p)
+
+    def forward(self, samples, sample_rate: int):
+        if len(samples) == 0:
+            warnings.warn(
+                "An empty samples tensor was passed to {}".format(
+                    self.__class__.__name__
+                )
+            )
+            return samples
+
+        if is_multichannel(samples):
+            if samples.shape[1] > samples.shape[2]:
+                warnings.warn(
+                    "Multichannel audio must have channels first, not channels last. In"
+                    " other words, the shape must be (batch size, channels, samples), not"
+                    " (batch_size, samples, channels)"
+                )
+            if not self.supports_multichannel:
+                raise MultichannelAudioNotSupportedException(
+                    "{} only supports mono audio, not multichannel audio".format(
+                        self.__class__.__name__
                     )
+                )
 
-            return self.apply(samples, sample_rate)
+        if not self.are_parameters_frozen:
+            batch_size = samples.shape[0]
+            self.parameters = {
+                "should_apply": self.bernoulli_distribution.sample(
+                    sample_shape=(batch_size,)
+                ).to(torch.bool)
+            }
+            if self.parameters["should_apply"].any():
+                self.randomize_parameters(
+                    samples[self.parameters["should_apply"]], sample_rate
+                )
+
+        if self.parameters["should_apply"].any():
+            cloned_samples = samples.detach().clone()
```

Comment:
Are we sure we want to detach here?

Reply:
Not sure, but this seemed to be the recommended way of cloning a tensor in pytorch
https://stackoverflow.com/questions/55266154/pytorch-preferred-way-to-copy-a-tensor

I thought I needed to clone it to avoid changing the original tensor, ref discussion on Slack

Reply:
But this removes the graph. 
`kornia` advertises for differentiability, I don't know what's their main use case, but if we do it in PyTorch, we might want to keep things backpropagable as well. 
The first use case I think about is a TasNet-like architecture with some kind of SpecAugment on the learned features, you won't want to detach in this case.

Reply:
Yeah okay, that's interesting. I guess I could remove the detach and keep the clone, then. Btw, how do we test (as in unit test) differentiability and check that the graph is kept intact?

Reply:
Yes.

For differentiability, you can backward anything that depends on the input and check that the gradients are not `None` anymore (with `requires_grad` inputs). Regarding the graph, I wouldn't unit test that, it's pytorch's job to handle it. You cannot change the graph on nodes "before you" anyway. 

We can use `with torch.autograd.detect_anomaly():` to make sure there is no NaN either.